2. Add Workflow Manager and Main

nikhitashanker77 opened the pull request:
Add conversion of JSON into workflow and create main and workflow manager classes.

- next PR will include adding logger in all cases of log.fatal & fmt.Println

nikhitashanker77 committed:
Add a mapping between names  and nodes.

nikhitashanker77 committed:
Add JSON to workflow creation.

nikhitashanker77 committed:
Add logger to main.

nikhitashanker77 committed:
remove use of public name

nikhitashanker77 committed:
fixes

nikhitashanker77 committed:
add metadata and args to json parsing

nikhitashanker77 committed:
Remove option 2

nikhitashanker commented on main.go:
```diff
@@ -0,0 +1,42 @@
+	var workflowManager WorkflowManager = newWorkflowManager(workflowsJson)
+
+	file, err := os.OpenFile(logging_file, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
+    if err != nil {
```
spacing on my IDE is fine but weird here.

MichaelBehm commented on workflows/workflow.go:
```diff
@@ -8,6 +8,7 @@ import (
 type InterfaceWorkflow interface {
 	AddNode(node nodes.InterfaceNode)
 	RunVirt()
```
Let's add the `Run()` function to workflow too and make `RunVirt()` private (`runVirt()`)

`Run()` should be defined in Workflow, and `Run()` will call `runVirt()`. The reason for this is if we want to do some logging for every single workflow we run, we don't want every workflow to have to recode all the logging.

Main/WorkflowManager should call `run()` not `runVirt()`

nikhitashanker77 replied:
I see done - used the abstract class approach previously discussed.

MichaelBehm commented on nodes/node.go:
```diff
@@ -6,6 +6,9 @@ import (
+	GetName() string
+	SetName(name string)
+	InitializeFields()
  	ProcessVirt(predecessorNodeResults []interface{})
```
see my "runVirt()" comment for Workflow and apply the same here for `Process()` and `processVirt()`

nikhitashanker replied:
Done

MichaelBehm commented on workflow_manager.go:
```diff
@@ -0,0 +1,57 @@
+func (workflowManager *WorkflowManager) constructWorkflow(workflowName string) workflows.InterfaceWorkflow {
+	var currWorkflowDescription WorkflowDescription
+	currWorkflowDescription = workflowManager.nameToWorkflow[workflowName]
+	if currWorkflowDescription.Type != "simple" {
```
let's make this a case-switch statement like we have to the NodeRegistry

nikhitashanker77 replied:
I see, done

MichaelBehm commented on workflow_manager.go:
```diff
@@ -0,0 +1,57 @@
+)
+
+/** Class for parsing workflow definitions. */
+type WorkflowManager struct {
```
The more I look at WorkflowManager the more I don't like it's name. Let's rename it to `WorkflowFactory` because that's basically what it is

nikhitashanker replied:
Done

MichaelBehm commented on workflows/simple_workflow.go:
```diff
@@ -20,4 +20,13 @@ func (simpleWorkflow *SimpleWorkflow) RunVirt() {
+func (simpleWorkflow *SimpleWorkflow) GetResult() map[string]interface{} {
+	var nodeToResult map[string]interface{} = make(map[string]interface{})
+	for _, node := range simpleWorkflow.Nodes {
+		nodeToResult[node.GetName()] = node.GetResult()
```
This is dangerous, because it's possible to have the same node appear multiple times in the same workflow. They'd have the same node.name, right?

nikhitashanker77 replied:
The python version only stores node results for the purpose of giving node n the results of nodes 0 to n-1 (getPredecessorResults) so I think I do not need this function. Removed GetResult in the Workflow class.

nikhitashanker replied:
I think getting the results of all nodes in a workflow is needed for testing (as in the Python version the only function that calls a workflow's getResult is the getPredecessorResults function).  I added a GetNodes function in the workflow just to make testing to see node's results possible (it would be nice if there was an @visiblefortesting in golang).

nikhitashanker77 replied:
I added a GetNodes function in the Workflow class to make seeing node's results possible. This would be used for testing purposes only (it would be nice if there was a JAVA @VisibleForTesting in golang).

nikhitashanker77 replied:
The  GetNodes function in the Workflow class can also be used to get the nodes results as we do in main.go.

MichaelBehm commented on node_registry.go:
```diff
@@ -0,0 +1,18 @@
+	"github.com/microsoft/BladeMonRT/nodes"
+)
+
+func makeInstance(typeName string) nodes.InterfaceNode {
```
Let's make this into a class and call it "NodeFactory"

Let `NodeFactory` have a public function `ConstructNode` given the string name of the node (basically just a copy of makeInstance here

MichaelBehm replied:
The WorkflowFactory can call this function

nikhitashanker77 replied:
Makes sense. I'm hesistant to create a class for this because this will be an empty struct NodeFactory. I changed the file name to node factory.

nikhitashanker77 replied:
I think there's no harm in creating an empty struct. Done.

MichaelBehm reviewed (changes requested):
See my comments

MichaelBehm commented on nodes/node.go:
```diff
@@ -6,6 +6,9 @@ import (
+	GetName() string
+	SetName(name string)
+	InitializeFields()
  	ProcessVirt(predecessorNodeResults []interface{})
```
nit: Rename `predecessorNodeResults` to `predecessorNodesResults`

nikhitashanker77 replied:
makes sense - done

nikhitashanker77 committed:
Fixes from code review.

nikhitashanker77 committed:
comments

nikhitashanker77 committed:
tests

nikhitashanker77 committed:
tests

MichaelBehm commented on workflow_factory.go:
```diff
@@ -0,0 +1,61 @@
+	Args interface{} `json:"args"`
+}
+
+func newWorkflowFactory(workflowsJson []byte) WorkflowFactory {
```
Make this capitalized as it is a public function, make this change for `constructWorkflow` to

nikhitashanker replied:
It's very specific to GO but capitalization at the start of a function name means that we want the function to be exported so it can be used in other packages. Since 'newWorkflowFactory' is only used within this package, lowercase n is fine here I think.

nikhitashanker77 replied:
It's very specific to GO but capitalization at the start of a function name means that we want the function to be exported so it can be used in other packages. Since 'newWorkflowFactory' is only used within this package, lowercase n is fine here I think.

MichaelBehm commented on workflow_factory.go:
```diff
@@ -0,0 +1,61 @@
+		
+	} 
+
+	// Create a collection of nodes from their names.
```
Move this above the switch

nikhitashanker77 replied:
done

MichaelBehm commented on workflow_factory.go:
```diff
@@ -0,0 +1,61 @@
+		workflowNodes = append(workflowNodes, node)
+	}
+
+	// Create a simple workflow and add nodes to the workflow.
```
Put this in the case "simple": bit

nikhitashanker77 replied:
I see, done.

MichaelBehm reviewed (changes requested):
See my comments

nikhitashanker77 commented:
In addition to addressing comments, I have created a "workflow context" and removed storing results at the nodes so that results are stored in one place. GO unfortunately has strong rules against package a importing package b and package b importing package a. Workaround will make the code structure hard to understand (it's sort of a hack like the abstract classes).

nikhitashanker77 committed:
fixes from code review.

nikhitashanker77 commented:
There was a problem where (because of https://github.com/golang/go/issues/23440#issuecomment-357476298) where if 2 dummynodes are created the 2 pointers to these two had the same memory address. This is as dangerous as having the same string map to 2 dummynodes, since we are using a map from pointer to interface{} to store the result. To fix this I added a dummyVariable of type interface in the Node struct.

nikhitashanker77 committed:
constructors

nikhitashanker77 committed:
format

nikhitashanker77 committed:
cleanup

nikhitashanker77 committed:
style

MichaelBehm commented on nodes/node.go:
```diff
@@ -4,24 +4,41 @@ import (
-// Interface for defining unit of work to be processed by event loop
+type WorkflowContext struct {
+	predecessorNodesResults []interface{}
+	nodeToResult map[InterfaceNode]interface{}
```
Here we're using 2 data structures to maintain a single state: the predecessor results

What would happen if a DummyNode instance called SaveResult multiple times? It will replace it's entry in `nodeToResult` yet append a new entry to `predecessorNodesResults`

I think it *should* overwrite whatever old result was written before without changing the length of `predecessorNodesResults`

The way Python BRT solves this issues is like so:
- On a per-workflow basis (in Workflow or WorkflowContext), keep an array of Node objects `nodes` which can be created upon the workflow's initialization
- Keep a `nodeToResult` map which maps Node objects to their results on a per-workflow basis
- When you want to `GetResults`, iterate over `nodes` and for each node, lookup it's results from the `nodeToResult` map
- `SetResults` is simply workflowContext.nodeToResult[this] = thisResults

MichaelBehm replied:
This way we can maintain an "order" of predecessors Nodes and their results without weird edge cases

nikhitashanker77 replied:
I see, thanks for the explanation. I have followed this pattern as closely as possible.

MichaelBehm reviewed (changes requested):
See my comments

nikhitashanker77 committed:
fixes from code review.

nikhitashanker77 committed:
nits

nikhitashanker77 committed:
Add a 'getPredecessorNodes' method.

nikhitashanker77 committed:
spacing

nikhitashanker77 committed:
spacing

nikhitashanker77 committed:
naming

nikhitashanker77 committed:
small fix:

nikhitashanker77 committed:
Change use of workflow context.

nikhitashanker77 committed:
Change use of workflow context.

MichaelBehm reviewed (approved):
Looks good, thanks!
